### ISSUE TITLE
[codegen] generated inplace/out= wrappers don't have input checks

### ISSUE THREAD

**Issue by bdhirsh** [pytorch/pytorch#58765]
I realized that after https://github.com/pytorch/xla/pull/2953 and https://github.com/pytorch/pytorch/pull/57510 landed, which codegen's `inplace` and `out=` wrappers for `functional` kernels, we're losing any input checking that was hardcoded into the inplace/functional kernel.

This isn't a huge problem, since ideally the functional kernel should be doing all of the input checking (using `at::meta::{op}` if the op is structured!).

The problem, though, is that there are some additional checks that inplace/out= kernels do, that functional kernels do not. The shape checks for inplace/out kernels are a little more comprehensive, since they include device and type promotion checks against the `out` or `self` tensor. 

We just need to update the generated inplace/out kernels to include these additional checks (plus ideally device checks for out= kernels, although I don't think XLA out kernels technically ervery did this:
```
at::Tensor & wrapper_Tensor_add_(at::Tensor & self, const at::Tensor & other, const at::Scalar & alpha) {
  // TODO: add a type promotion check here
  // at::ScalarType resultType = at::result_type(self, other);
  // TORCH_CHECK(at::canCast(/*from=*/resultType, /*to=*/self.scalar_type()));
  auto wrapper_Tensor_add__tmp = wrapper_Tensor_add(self, other, alpha);
  at::_copy_from(wrapper_Tensor_add__tmp, self);
  return self;
}
```

These checks are already covered by TensorIterator, so it would be nice to factor them in some way and make them reusable, but that's not a strict requirement.

**Comment by ezyang** [comment]
I promised @ngimel that I would do this, just haven't had time yet

**Comment by bdhirsh** [comment]
> I promised @ngimel that I would do this, just haven't had time yet

Hmmm the problem that I meant to bring up in this issue is specific to XLA's usage of the generated inplace/out wrappers. Are you referring to other places in the codegen where in-tree kernels are missing inplace/out= checks?

Maybe that applies to non-TensorIterator kernels, since it's pretty hard to enforce that every kernel writer who doesn't use TensorIterator does the extra `self`/`out` checks, and it would make more sense to centralize it through codegen (like Wenlei did for device checks).

**Comment by bdhirsh** [comment]
It looks like structured kernels [already get some codegen'd type promotion checks](https://github.com/pytorch/pytorch/blob/444e195b6da6eefc2b191af0e1aaaf2d1501fe65/tools/codegen/dest/register_dispatch_key.py#L381) for the inplace/out= wrappers, as long as the op's `meta()` function correctly calculated the common dtype. Are you proposing codegen'ing some of that logic to calculate common dtype?

**Comment by ezyang** [comment]
Yeah, it's specifically non-TensorIterator, need to make sure we don't try to broadcast self in inplace situation.
